File: decompiled-improved/io/legado/app/model/localBook/CbzFile.java
/*
 * Decompiled with CFR 0.152.
 * 
 * Could not load the following classes:
 *  kotlin.Metadata
 *  kotlin.Pair
 *  kotlin.collections.CollectionsKt
 *  kotlin.jvm.internal.DefaultConstructorMarker
 *  kotlin.jvm.internal.Intrinsics
 *  kotlin.text.StringsKt
 *  org.jetbrains.annotations.NotNull
 *  org.jetbrains.annotations.Nullable
 */
package io.legado.app.model.localBook;

import io.legado.app.data.entities.Book;
import io.legado.app.data.entities.BookChapter;
import io.legado.app.utils.FileUtils;
import io.legado.app.utils.MD5Utils;
import io.legado.app.utils.XmlUtils;
import java.io.File;
import java.io.InputStream;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.Enumeration;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.zip.ZipEntry;
import java.util.zip.ZipFile;
import kotlin.Metadata;
import kotlin.Pair;
import kotlin.collections.CollectionsKt;
import kotlin.jvm.internal.DefaultConstructorMarker;
import kotlin.jvm.internal.Intrinsics;
import kotlin.text.StringsKt;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

@Metadata(mv={1, 5, 1}, k=1, xi=48, d1={"\u0000@\n\u0002\u0018\u0002\n\u0002\u0010\u0000\n\u0000\n\u0002\u0018\u0002\n\u0002\b\u0005\n\u0002\u0018\u0002\n\u0002\b\u0005\n\u0002\u0010%\n\u0002\u0010\u000e\n\u0002\b\u0005\n\u0002\u0018\u0002\n\u0002\u0018\u0002\n\u0002\b\u0003\n\u0002\u0018\u0002\n\u0000\n\u0002\u0010\u0002\n\u0002\b\u0003\u0018\u0000 \u001f2\u00020\u0001:\u0001\u001fB\r\u0012\u0006\u0010\u0002\u001a\u00020\u0003\u00a2\u0006\u0002\u0010\u0004J\u000e\u0010\u0015\u001a\b\u0012\u0004\u0012\u00020\u00170\u0016H\u0002J\u0012\u0010\u0018\u001a\u0004\u0018\u00010\u00102\u0006\u0010\u0019\u001a\u00020\u0017H\u0002J$\u0010\u001a\u001a\u001e\u0012\u0012\u0012\u0010\u0012\u0004\u0012\u00020\u0010\u0012\u0004\u0012\u00020\u0001\u0018\u00010\u000f\u0012\u0006\u0012\u0004\u0018\u00010\t0\u001bH\u0002J\b\u0010\u001c\u001a\u00020\u001dH\u0002J\b\u0010\u001e\u001a\u00020\u001dH\u0002R\u001a\u0010\u0002\u001a\u00020\u0003X\u0086\u000e\u00a2\u0006\u000e\n\u0000\u001a\u0004\b\u0005\u0010\u0006\"\u0004\b\u0007\u0010\u0004R\u001c\u0010\b\u001a\u0004\u0018\u00010\tX\u0086\u000e\u00a2\u0006\u000e\n\u0000\u001a\u0004\b\n\u0010\u000b\"\u0004\b\f\u0010\rR(\u0010\u000e\u001a\u0010\u0012\u0004\u0012\u00020\u0010\u0012\u0004\u0012\u00020\u0001\u0018\u00010\u000fX\u0086\u000e\u00a2\u0006\u000e\n\u0000\u001a\u0004\b\u0011\u0010\u0012\"\u0004\b\u0013\u0010\u0014\u00a8\u0006 "}, d2={"Lio/legado/app/model/localBook/CbzFile;", "", "book", "Lio/legado/app/data/entities/Book;", "(Lio/legado/app/data/entities/Book;)V", "getBook", "()Lio/legado/app/data/entities/Book;", "setBook", "cover", "Ljava/io/InputStream;", "getCover", "()Ljava/io/InputStream;", "setCover", "(Ljava/io/InputStream;)V", "info", "", "", "getInfo", "()Ljava/util/Map;", "setInfo", "(Ljava/util/Map;)V", "getChapterList", "Ljava/util/ArrayList;", "Lio/legado/app/data/entities/BookChapter;", "getContent", "chapter", "parseBookInfo", "Lkotlin/Pair;", "upBookInfo", "", "updateCover", "Companion", "reader-pro"})
public final class CbzFile {
    @NotNull
    public static final Companion Companion = new Companion(null);
    @NotNull
    private Book book;
    @Nullable
    private Map<String, Object> info;
    @Nullable
    private InputStream cover;
    @Nullable
    private static CbzFile cFile;

    public CbzFile(@NotNull Book book) {
        Intrinsics.checkNotNullParameter((Object)book, (String)"book");
        this.book = book;
    }

    @NotNull
    public final Book getBook() {
        return this.book;
    }

    public final void setBook(@NotNull Book book) {
        Intrinsics.checkNotNullParameter((Object)book, (String)"<set-?>");
        this.book = book;
    }

    @Nullable
    public final Map<String, Object> getInfo() {
        return this.info;
    }

    public final void setInfo(@Nullable Map<String, Object> map) {
        this.info = map;
    }

    @Nullable
    public final InputStream getCover() {
        return this.cover;
    }

    public final void setCover(@Nullable InputStream inputStream) {
        this.cover = inputStream;
    }

    private final Pair<Map<String, Object>, InputStream> parseBookInfo() {
        if (this.cover != null || this.info != null) {
            return new Pair(this.info, (Object)this.cover);
        }
        ZipFile zf = new ZipFile(this.book.getLocalFile());
        Enumeration<? extends ZipEntry> entries = zf.entries();
        Object[] objectArray = new String[]{"jpg", "jpeg", "gif", "png", "bmp", "webp", "svg"};
        List imageExt = CollectionsKt.listOf((Object[])objectArray);
        while (entries.hasMoreElements()) {
            ZipEntry zipEntry = entries.nextElement();
            if (zipEntry == null) {
                throw new NullPointerException("null cannot be cast to non-null type java.util.zip.ZipEntry");
            }
            ZipEntry zipEntry2 = zipEntry;
            if (!zipEntry2.isDirectory()) {
                Object object;
                String name = zipEntry2.getName();
                if (name.equals("ComicInfo.xml")) {
                    InputStream inputStream = zf.getInputStream(zipEntry2);
                    object = inputStream;
                    Intrinsics.checkNotNullExpressionValue((Object)object, (String)"inputStream");
                    this.info = XmlUtils.INSTANCE.xml2map(object);
                } else if (this.cover == null) {
                    String ext;
                    Intrinsics.checkNotNullExpressionValue((Object)name, (String)"name");
                    object = FileUtils.getFileExtetion$default(FileUtils.INSTANCE, name, null, 2, null);
                    boolean bl = false;
                    Object object2 = object;
                    if (object2 == null) {
                        throw new NullPointerException("null cannot be cast to non-null type java.lang.String");
                    }
                    Intrinsics.checkNotNullExpressionValue((Object)((String)object2).toLowerCase(Locale.ROOT), (String)"(this as java.lang.Strin\u2026.toLowerCase(Locale.ROOT)");
                    if (imageExt.contains(ext)) {
                        this.cover = zf.getInputStream(zipEntry2);
                    }
                }
            }
            if (this.cover == null || this.info == null) continue;
        }
        return new Pair(this.info, (Object)this.cover);
    }

    private final void upBookInfo() {
        Pair<Map<String, Object>, InputStream> result2 = this.parseBookInfo();
        if (result2.getFirst() != null) {
            Map info;
            Object object = result2.getFirst();
            if (object == null) {
                throw new NullPointerException("null cannot be cast to non-null type kotlin.collections.Map<kotlin.String, kotlin.Any>");
            }
            Map bookInfo = (Map)object;
            Map map = (Map)bookInfo.get("ComicInfo");
            Map map2 = info = map == null ? null : map;
            map = map2 == null ? null : map2.get("Title");
            this.book.setName((String)(map == null ? this.book.getName() : map));
            map2 = info;
            map = map2 == null ? null : map2.get("Writer");
            this.book.setAuthor((String)(map == null ? this.book.getAuthor() : map));
        }
        this.updateCover();
    }

    private final void updateCover() {
        Pair<Map<String, Object>, InputStream> result2;
        String coverFile = Intrinsics.stringPlus((String)MD5Utils.INSTANCE.md5Encode16(this.book.getBookUrl()), (Object)".jpg");
        String[] stringArray = new String[]{this.book.getUserNameSpace(), "covers", coverFile};
        String relativeCoverUrl = ((Object)Paths.get("assets", stringArray)).toString();
        this.book.setCoverUrl(Intrinsics.stringPlus((String)"/", (Object)StringsKt.replace$default((String)relativeCoverUrl, (String)"\\", (String)"/", (boolean)false, (int)4, null)));
        String[] stringArray2 = new String[]{"storage", relativeCoverUrl};
        String coverUrl = ((Object)Paths.get(this.book.workRoot(), stringArray2)).toString();
        if (!new File(coverUrl).exists() && (result2 = this.parseBookInfo()).getSecond() != null) {
            Object object = result2.getSecond();
            if (object == null) {
                throw new NullPointerException("null cannot be cast to non-null type java.io.InputStream");
            }
            InputStream coverStream = (InputStream)object;
            FileUtils.INSTANCE.writeInputStream(coverUrl, coverStream);
        }
    }

    private final String getContent(BookChapter chapter) {
        return "";
    }

    private final ArrayList<BookChapter> getChapterList() {
        BookChapter bookChapter;
        ArrayList<BookChapter> chapterList = new ArrayList<BookChapter>();
        ZipFile zf = new ZipFile(this.book.getLocalFile());
        Enumeration<? extends ZipEntry> entries = zf.entries();
        int n = 0;
        ArrayList<String> imageFileList = new ArrayList<String>();
        while (entries.hasMoreElements()) {
            ZipEntry zipEntry = entries.nextElement();
            if (zipEntry == null) {
                throw new NullPointerException("null cannot be cast to non-null type java.util.zip.ZipEntry");
            }
            ZipEntry zipEntry2 = zipEntry;
            if (zipEntry2.isDirectory()) continue;
            String name = zipEntry2.getName();
            Intrinsics.checkNotNullExpressionValue((Object)name, (String)"name");
            if (StringsKt.endsWith$default((String)name, (String)".xml", (boolean)false, (int)2, null)) continue;
            imageFileList.add(name);
        }
        CollectionsKt.sort((List)imageFileList);
        n = 0;
        int n2 = imageFileList.size();
        if (n < n2) {
            do {
                int i = n++;
                Object e = imageFileList.get(i);
                Intrinsics.checkNotNullExpressionValue(e, (String)"imageFileList.get(i)");
                String name = (String)e;
                BookChapter chapter = new BookChapter(null, null, false, null, null, 0, null, null, null, null, null, null, null, 8191, null);
                chapter.setTitle(name);
                chapter.setIndex(i);
                chapter.setBookUrl(this.book.getBookUrl());
                chapter.setUrl(name);
                chapterList.add(chapter);
            } while (n < n2);
        }
        this.book.setLatestChapterTitle((bookChapter = (BookChapter)CollectionsKt.lastOrNull((List)chapterList)) == null ? null : bookChapter.getTitle());
        this.book.setTotalChapterNum(chapterList.size());
        return chapterList;
    }

    @Metadata(mv={1, 5, 1}, k=1, xi=48, d1={"\u00008\n\u0002\u0018\u0002\n\u0002\u0010\u0000\n\u0002\b\u0002\n\u0002\u0018\u0002\n\u0002\b\u0002\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0002\u0018\u0002\n\u0000\n\u0002\u0010\u000e\n\u0002\b\u0002\n\u0002\u0010\u0002\n\u0000\n\u0002\u0010\u000b\n\u0000\b\u0086\u0003\u0018\u00002\u00020\u0001B\u0007\b\u0002\u00a2\u0006\u0002\u0010\u0002J\u0010\u0010\u0005\u001a\u00020\u00042\u0006\u0010\u0006\u001a\u00020\u0007H\u0002J\u0014\u0010\b\u001a\b\u0012\u0004\u0012\u00020\n0\t2\u0006\u0010\u0006\u001a\u00020\u0007J\u0018\u0010\u000b\u001a\u0004\u0018\u00010\f2\u0006\u0010\u0006\u001a\u00020\u00072\u0006\u0010\r\u001a\u00020\nJ\u0018\u0010\u000e\u001a\u00020\u000f2\u0006\u0010\u0006\u001a\u00020\u00072\b\b\u0002\u0010\u0010\u001a\u00020\u0011R\u0010\u0010\u0003\u001a\u0004\u0018\u00010\u0004X\u0082\u000e\u00a2\u0006\u0002\n\u0000\u00a8\u0006\u0012"}, d2={"Lio/legado/app/model/localBook/CbzFile$Companion;", "", "()V", "cFile", "Lio/legado/app/model/localBook/CbzFile;", "getCbzFile", "book", "Lio/legado/app/data/entities/Book;", "getChapterList", "Ljava/util/ArrayList;", "Lio/legado/app/data/entities/BookChapter;", "getContent", "", "chapter", "upBookInfo", "", "onlyCover", "", "reader-pro"})
    public static final class Companion {
        private Companion() {
        }

        private final synchronized CbzFile getCbzFile(Book book) {
            CbzFile cbzFile;
            block7: {
                block6: {
                    String string;
                    if (cFile == null) break block6;
                    cbzFile = cFile;
                    if (cbzFile == null) {
                        string = null;
                    } else {
                        Book book2 = cbzFile.getBook();
                        string = book2.getBookUrl();
                    }
                    if (Intrinsics.areEqual(string, (Object)book.getBookUrl())) break block7;
                }
                cFile = new CbzFile(book);
                CbzFile cbzFile2 = cFile;
                Intrinsics.checkNotNull((Object)cbzFile2);
                return cbzFile2;
            }
            cbzFile = cFile;
            if (cbzFile != null) {
                cbzFile.setBook(book);
            }
            CbzFile cbzFile3 = cFile;
            Intrinsics.checkNotNull((Object)cbzFile3);
            return cbzFile3;
        }

        @NotNull
        public final synchronized ArrayList<BookChapter> getChapterList(@NotNull Book book) {
            Intrinsics.checkNotNullParameter((Object)book, (String)"book");
            return this.getCbzFile(book).getChapterList();
        }

        @Nullable
        public final synchronized String getContent(@NotNull Book book, @NotNull BookChapter chapter) {
            Intrinsics.checkNotNullParameter((Object)book, (String)"book");
            Intrinsics.checkNotNullParameter((Object)chapter, (String)"chapter");
            return this.getCbzFile(book).getContent(chapter);
        }

        public final synchronized void upBookInfo(@NotNull Book book, boolean onlyCover) {
            Intrinsics.checkNotNullParameter((Object)book, (String)"book");
            if (onlyCover) {
                this.getCbzFile(book).updateCover();
                return;
            }
            this.getCbzFile(book).upBookInfo();
        }

        public static /* synthetic */ void upBookInfo$default(Companion companion, Book book, boolean bl, int n, Object object) {
            if ((n & 2) != 0) {
                bl = false;
            }
            companion.upBookInfo(book, bl);
        }

        public /* synthetic */ Companion(DefaultConstructorMarker $constructor_marker) {
            this();
        }
    }
}

File: full-vineflower/io/legado/app/model/localBook/CbzFile.kt
package io.legado.app.model.localBook

import io.legado.app.data.entities.Book
import io.legado.app.data.entities.BookChapter
import io.legado.app.utils.FileUtils
import io.legado.app.utils.MD5Utils
import io.legado.app.utils.XmlUtils
import java.io.File
import java.io.InputStream
import java.nio.file.Paths
import java.util.ArrayList
import java.util.Enumeration
import java.util.Locale
import java.util.zip.ZipEntry
import java.util.zip.ZipFile
import kotlin.jvm.internal.Intrinsics

public class CbzFile(book: Book) {
   public final var book: Book
      internal set

   public final var cover: InputStream?
      internal set

   public final var info: MutableMap<String, Any>?
      internal set

   init {
      this.book = book;
   }

   private fun parseBookInfo(): Pair<MutableMap<String, Any>?, InputStream?> {
      if (this.cover == null && this.info == null) {
         val zf: ZipFile = new ZipFile(this.book.getLocalFile());
         val entries: Enumeration = zf.entries();
         val imageExt: java.util.List = CollectionsKt.listOf(new java.lang.String[]{"jpg", "jpeg", "gif", "png", "bmp", "webp", "svg"});

         while (entries.hasMoreElements()) {
            var name: java.lang.String = (java.lang.String)entries.nextElement();
            if (name == null) {
               throw new NullPointerException("null cannot be cast to non-null type java.util.zip.ZipEntry");
            }

            val var9: ZipEntry = name as ZipEntry;
            if (!(name as ZipEntry).isDirectory()) {
               name = var9.getName();
               if (name.equals("ComicInfo.xml")) {
                  val ext: InputStream = zf.getInputStream(var9);
                  val var10001: XmlUtils = XmlUtils.INSTANCE;
                  this.info = var10001.xml2map(ext);
               } else if (this.cover == null) {
                  val var10000: FileUtils = FileUtils.INSTANCE;
                  val var7: java.lang.String = FileUtils.getFileExtetion$default(var10000, name, null, 2, null);
                  if (var7 == null) {
                     throw new NullPointerException("null cannot be cast to non-null type java.lang.String");
                  }

                  val var12: java.lang.String = var7.toLowerCase(Locale.ROOT);
                  if (imageExt.contains(var12)) {
                     this.cover = zf.getInputStream(var9);
                  }
               }
            }

            if (this.cover != null && this.info != null) {
               break;
            }
         }

         return new Pair<>(this.info, this.cover);
      } else {
         return new Pair<>(this.info, this.cover);
      }
   }

   private fun upBookInfo() {
      val result: Pair = this.parseBookInfo();
      if (result.getFirst() != null) {
         var info: java.util.Map = (java.util.Map)result.getFirst();
         if (info == null) {
            throw new NullPointerException("null cannot be cast to non-null type kotlin.collections.Map<kotlin.String, kotlin.Any>");
         }

         var var4: java.util.Map = info.get("ComicInfo") as java.util.Map;
         info = if (var4 == null) null else var4;
         var4 = (java.util.Map)(if ((if (var4 == null) null else var4) == null) null else (if (var4 == null) null else var4).get("Title"));
         this.book.setName((if (var4 == null) this.book.getName() else var4) as java.lang.String);
         var4 = (java.util.Map)(if (info == null) null else info.get("Writer"));
         this.book.setAuthor((if (var4 == null) this.book.getAuthor() else var4) as java.lang.String);
      }

      this.updateCover();
   }

   private fun updateCover() {
      val relativeCoverUrl: java.lang.String = Paths.get(
            "assets", this.book.getUserNameSpace(), "covers", Intrinsics.stringPlus(MD5Utils.INSTANCE.md5Encode16(this.book.getBookUrl()), ".jpg")
         )
         .toString();
      this.book.setCoverUrl(Intrinsics.stringPlus("/", StringsKt.replace$default(relativeCoverUrl, "\\", "/", false, 4, null)));
      val var7: java.lang.String = Paths.get(this.book.workRoot(), "storage", relativeCoverUrl).toString();
      if (!new File(var7).exists()) {
         val var8: Pair = this.parseBookInfo();
         if (var8.getSecond() != null) {
            val var6: Any = var8.getSecond();
            if (var6 == null) {
               throw new NullPointerException("null cannot be cast to non-null type java.io.InputStream");
            }

            FileUtils.INSTANCE.writeInputStream(var7, var6 as InputStream);
         }
      }
   }

   private fun getContent(chapter: BookChapter): String? {
      return "";
   }

   private fun getChapterList(): ArrayList<BookChapter> {
      val chapterList: ArrayList = new ArrayList();
      val entries: Enumeration = new ZipFile(this.book.getLocalFile()).entries();
      val imageFileList: ArrayList = new ArrayList();

      while (entries.hasMoreElements()) {
         var name: java.lang.String = (java.lang.String)entries.nextElement();
         if (name == null) {
            throw new NullPointerException("null cannot be cast to non-null type java.util.zip.ZipEntry");
         }

         val var10: ZipEntry = name as ZipEntry;
         if (!(name as ZipEntry).isDirectory()) {
            name = var10.getName();
            if (!StringsKt.endsWith$default(name, ".xml", false, 2, null)) {
               imageFileList.add(name);
            }
         }
      }

      CollectionsKt.sort(imageFileList);
      var var11: Int = 0;
      val var14: Int = imageFileList.size();
      if (0 < var14) {
         do {
            val i: Int = var11++;
            var chapter: BookChapter = (BookChapter)imageFileList.get(i);
            val namex: java.lang.String = chapter as java.lang.String;
            chapter = new BookChapter(null, null, false, null, null, 0, null, null, null, null, null, null, null, 8191, null);
            chapter.setTitle(namex);
            chapter.setIndex(i);
            chapter.setBookUrl(this.book.getBookUrl());
            chapter.setUrl(namex);
            chapterList.add(chapter);
         } while (var11 < var14);
      }

      val var10000: Book = this.book;
      val var12: BookChapter = CollectionsKt.lastOrNull(chapterList);
      var10000.setLatestChapterTitle(if (var12 == null) null else var12.getTitle());
      this.book.setTotalChapterNum(chapterList.size());
      return chapterList;
   }

   public companion object {
      private final var cFile: CbzFile?

      @Synchronized
      private fun getCbzFile(book: Book): CbzFile {
         if (CbzFile.access$getCFile$cp() != null) {
            var var2: CbzFile = CbzFile.access$getCFile$cp();
            if ((if (var2 == null) null else var2.getBook().getBookUrl()) == book.getBookUrl()) {
               var2 = CbzFile.access$getCFile$cp();
               if (var2 != null) {
                  var2.setBook(book);
               }

               val var6: CbzFile = CbzFile.access$getCFile$cp();
               return var6;
            }
         }

         CbzFile.access$setCFile$cp(new CbzFile(book));
         val var5: CbzFile = CbzFile.access$getCFile$cp();
         return var5;
      }

      @Synchronized
      public fun getChapterList(book: Book): ArrayList<BookChapter> {
         return CbzFile.access$getChapterList(this.getCbzFile(book));
      }

      @Synchronized
      public fun getContent(book: Book, chapter: BookChapter): String? {
         return CbzFile.access$getContent(this.getCbzFile(book), chapter);
      }

      @Synchronized
      public fun upBookInfo(book: Book, onlyCover: Boolean = false) {
         if (onlyCover) {
            CbzFile.access$updateCover(this.getCbzFile(book));
         } else {
            CbzFile.access$upBookInfo(this.getCbzFile(book));
         }
      }
   }
}
